### PR TITLE
refactor authorization provider setup

### PR DESCRIPTION
**Pull request by tomerd**
motivation: more easily share authorization provider setup with libSwiftPM users

changes:
* move authorization providers setup from SwiftTool to a Worksapce configuraiton utility
* adjust call sites and tests

### CODE REVIEW

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Review by neonichu — APPROVED** [review]

File: Sources/Workspace/WorkspaceConfiguration.swift
Comment (yim-lee) [inline Sources/Workspace/WorkspaceConfiguration.swift:240]: Please wait for the outcome of https://github.com/apple/swift-package-manager/pull/4075 before merging this

File: Sources/Workspace/WorkspaceConfiguration.swift
Comment (neonichu) [inline Sources/Workspace/WorkspaceConfiguration.swift:301]: Would it make sense to make `rootPath` optional here or split this up into two cases? I see the case for "default global" and "default global plus local at this root path" (this is what we are doing right now).

File: Sources/Commands/SwiftTool.swift
Comment (neonichu) [inline Sources/Commands/SwiftTool.swift:582]: I know this is just moving the code, but I am not sure if it this is actually correct for multi root. This will be a file path, not a directory, so essentially never work.

Comment (tomerd) [reply Sources/Commands/SwiftTool.swift:582]: interesting, the entire code in SwiftTool treats `multirootPackageDataFile` as the root directory, sounds like that's working (and that we are missing some tests?)

File: Sources/Workspace/WorkspaceConfiguration.swift
Comment (tomerd) [inline Sources/Workspace/WorkspaceConfiguration.swift:298]: @yim-lee this is my version of the fix in #4075, I would like to also add APIs to FileSystem to check for readability/writability so we dont have to hack around it with temp files, etc

**Comment by tomerd** [comment]
@swift-ci please smoke test

File: Sources/Workspace/WorkspaceConfiguration.swift
Comment (yim-lee) [reply Sources/Workspace/WorkspaceConfiguration.swift:298]: reference: rdar://88308095

File: Sources/Workspace/WorkspaceConfiguration.swift
Comment (yim-lee) [inline Sources/Workspace/WorkspaceConfiguration.swift:297]: Replicating Boris' suggestion:

```suggestion
                observabilityScope.emit(warning: "\(path.parentDirectory) is not accessible or not readable, not using .netrc file in it: \(error)")
```

File: Sources/Workspace/WorkspaceConfiguration.swift
Comment (yim-lee) [reply Sources/Workspace/WorkspaceConfiguration.swift:298]: We can't merge this PR without the FileSystem API then, correct? It's the writability that's problematic.

Comment (tomerd) [reply Sources/Workspace/WorkspaceConfiguration.swift:298]: no in this case it needs to be able to read, which this test is a good approximation for. I am planning to add the FS APIs in a separate PR as its a more involved change

File: Sources/Workspace/WorkspaceConfiguration.swift
Comment (tomerd) [reply Sources/Workspace/WorkspaceConfiguration.swift]: yes had the same thought and changed accordingly

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@swift-ci please smoke test

File: Sources/Workspace/WorkspaceConfiguration.swift
Comment (tomerd) [reply Sources/Workspace/WorkspaceConfiguration.swift:289]: @yim-lee this is now using the new readability API

Comment (yim-lee) [reply Sources/Workspace/WorkspaceConfiguration.swift:289]: See my comment in https://github.com/apple/swift-package-manager/pull/4079#discussion_r798045066. We need `isWritable` unless we decide to remove write capability from `NetrcAuthorizationProvider`.

File: Tests/WorkspaceTests/AuthorizationProviderTests.swift
Comment (yim-lee) [inline Tests/WorkspaceTests/AuthorizationProviderTests.swift:4]:
```suggestion
 Copyright (c) 2022 Apple Inc. and the Swift project authors
```

File: Sources/Commands/SwiftTool.swift
Comment (neonichu) [reply Sources/Commands/SwiftTool.swift:586]: Oh, I see how it works, `options.multirootPackageDataFile` is actually pointing to a `.xcworkspace` which is a directory. I think we should change this to use `xcshareddata` because this is a configuration file.

Comment (neonichu) [reply Sources/Commands/SwiftTool.swift:586]: To be clear, what I mean is use `multiRootPackageDataFile.appending(components: "xcshareddata")` as the root path here.

Comment (tomerd) [reply Sources/Commands/SwiftTool.swift:586]: sounds like good material for follow on PR? it is used in various spots

**Comment by tomerd** [comment]
lets get #4079 in first, then we can rebase this one on top

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@yim-lee @neonichu this is ready for review

**Review by yim-lee — APPROVED** [review]
We should probably run source compat since we've had problems with related code this week.

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
source compat looks good: https://github.com/apple/swift/pull/41142

i think this is good to go. @yim-lee want to take one last look?

**Comment by yim-lee** [comment]
LGTM 👍

**Pull request merged**
